README.md
Ionic-Cli
=========

The Ionic Framework command line utility makes it easy to start, build, run, and emulate [Ionic](http://ionicframework.com/) apps. In the future, it will also have support for our mobile development services and tools that make Ionic even more powerful.

Use the `ionic --help` command for more detailed task information.

## Installing

```bash
$ npm install -g ionic
```

*Note: For a global install of `-g ionic`, OSX/Linux users may need to prefix the command with `sudo`.*


## Starting an Ionic App

```bash
$ ionic start myapp [template]
```

Starter templates can either come from a named template, a Github repo, a Codepen, or a local directory. A starter template is what becomes the `www` directory within the Cordova project.

__Named template starters:__

* [tabs](https://github.com/driftyco/ionic-starter-tabs) (Default)
* [sidemenu](https://github.com/driftyco/ionic-starter-sidemenu)
* [blank](https://github.com/driftyco/ionic-starter-blank)

__Github Repo starters:__

* Any Github repo url, ex: [https://github.com/driftyco/ionic-starter-tabs](https://github.com/driftyco/ionic-starter-tabs)
* Named templates are simply aliases to Ionic starter repos

__Codepen URL starters:__

* Any Codepen url, ex: [http://codepen.io/ionic/pen/odqCz](http://codepen.io/ionic/pen/odqCz)
* [Ionic Codepen Demos](http://codepen.io/ionic/public-list/)

__Local directory starters:__

* Relative or absolute path to a local directory

__Command-line flags/options:__

    --appname, -a  .......  Human readable name for the app
                            (Use quotes around the name)
    --id, -i  ............  Package name set in the <widget id> config
                            ex: com.mycompany.myapp
    --no-cordova, -w  ....  Do not create an app targeted for Cordova


## Testing in a Browser

Use `ionic serve` to start a local development server for app dev and testing. This is useful for both desktop browser testing, and to test within a device browser which is connected to the same network. Additionally, this command starts LiveReload which is used to monitor changes in the file system. As soon as you save a file the browser is refreshed automatically. View [Using Sass](https://github.com/driftyco/ionic-cli/blob/master/README.md#using-sass) if you would also like to have `ionic serve` watch the project's Sass files.

```bash
$ ionic serve [options]
```

__Service Proxies:__

The `serve` command can add some proxies to the http server. These proxies are useful if you are developing in the browser and you need to make calls to an external API. With this feature you can proxy request to the external api through the ionic http server preventing the `No 'Access-Control-Allow-Origin' header is present on the requested resource` error.

In the `ionic.project` file you can add a property with an array of proxies you want to add. The proxies are object with two properties:

* `path`: string that will be matched against the beginning of the incoming request URL.
* `proxyUrl`: a string with the url of where the proxied request should go.

```json
{
  "name": "appname",
  "email": "",
  "app_id": "",
  "proxies": [
    {
      "path": "/v1",
      "proxyUrl": "https://api.instagram.com/v1"
    }
  ]
}

```

Using the above configuration, you can now make requests to your local server at `http://localhost:8100/v1` to have it proxy out requests to `https://api.instagram.com/v1`

For example:

```js
angular.module('starter.controllers', [])
.constant('InstagramApiUrl', '')
// .contant('InstagramApiUrl','https://api.instagram.com')
//In production, make this the real URL

.controller('FeedCtrl', function($scope, $http, InstagramApiUrl) {

  $scope.feed = null;

  $http.get(InstagramApiUrl + '/v1/media/search?client_id=1&lat=48&lng=2.294351').then(function(data) {
    console.log('data ' , data)
    $scope.feed = data;
  })

})
```

See also [this gist](https://gist.github.com/jbavari/d9c1c94058c4fdd4e935) for more help.

__Command-line flags/options:__

    [--consolelogs|-c] ......  Print app console logs to Ionic CLI
    [--serverlogs|-s] .......  Print dev server logs to Ionic CLI
    [--port|-p] .............  Dev server HTTP port (8100 default)
    [--livereload-port|-i] ..  Live Reload port (35729 default)
    [--nobrowser|-b] ........  Disable launching a browser
    [--nolivereload|-r] .....  Do not start live reload
    [--noproxy|-x] ..........  Do not add proxies


## Adding a platform target

```bash
$ ionic platform ios android
```

## Building your app

```bash
$ ionic build ios
```

## Live Reload App During Development (beta)

The `run` or `emulate` command will deploy the app to the specified platform devices/emulators. You can also run __live reload__ on the specified platform device by adding the `--livereload` option. The live reload functionality is similar to `ionic serve`, but instead of developing and debugging an app using a standard browser, the compiled hybrid app itself is watching for any changes to its files and reloading the app when needed. This reduces the requirement to constantly rebuild the app for small changes. However, any changes to plugins will still require a full rebuild. For live reload to work, the dev machine and device must be on the same local network, and the device must support [web sockets](http://caniuse.com/websockets).

With live reload enabled, an app's console logs can also be printed to the terminal/command prompt by including the `--consolelogs` or `-c` option. Additionally, the development server's request logs can be printed out using `--serverlogs` or `-s` options.

__Command-line flags/options for `run` and `emulate`:__

    [--livereload|-l] .......  Live Reload app dev files from the device (beta)
    [--consolelogs|-c] ......  Print app console logs to Ionic CLI (live reload req.)
    [--serverlogs|-s] .......  Print dev server logs to Ionic CLI (live reload req.)
    [--port|-p] .............  Dev server HTTP port (8100 default, live reload req.)
    [--livereload-port|-i] ..  Live Reload port (35729 default, live reload req.)
    [--debug|--release]

While the server is running for live reload, you can use the following commands within the CLI:

    restart or r to restart the client app from the root
    goto or g and a url to have the app navigate to the given url
    consolelogs or c to enable/disable console log output
    serverlogs or s to enable/disable server log output
    quit or q to shutdown the server and exit


## Emulating your app

Deploys the Ionic app on specified platform emulator. This is simply an alias for `run --emulator`.

```bash
$ ionic emulate ios [options]
```


## Running your app

Deploys the Ionic app on specified platform devices. If a device is not found it'll then deploy to an emulator/simulator.

```bash
$ ionic run ios [options]
```

## Icon and Splash Screen Image Generation

Automatically generate icons and splash screens from source images to create each size needed for each platform, in addition to copying each resized and cropped image into each platform's resources directory. Source images can either be a `png` or `psd` __Photoshop file__. Images are generated using Ionic's image resizing and cropping server, instead of requiring special libraries and plugins to be installed locally.

Since each platform has different image requirements, it's best to make a source image for the largest size needed, and let the CLI do all the resizing, cropping and copying for you. Newly generated images will be placed in the `resources` directory at the root of the Cordova project. Additionally, the CLI will update and add the correct `<platform>` configs to the project's [config.xml](http://cordova.apache.org/docs/en/edge/config_ref_images.md.html#Icons%20and%20Splash%20Screens) file.

During the build process, Cordova will look through the project's [config.xml](http://cordova.apache.org/docs/en/edge/config_ref_images.md.html#Icons%20and%20Splash%20Screens) file and copy the newly created resource images to the platform's specific resource folder. For example, Android's resource folder can be found in `platforms/android/res`, and iOS uses `platforms/ios/APP_NAME/Resources`.


### Icon Source Image

Save an `icon.png` or `icon.psd` file within the `resources` directory at the root of the Cordova project. The icon image's minimum dimensions should be 192x192, and should have __no__ rounded corners. Note that each platform will apply it's own mask and effects to the icons. For example, iOS will automatically apply it's custom rounded corners, so the source file should not already come with rounded corners.

```bash
$ ionic resources --icon
```


### Splash Screen Source Image

Save a `splash.png` or `splash.psd` file within the `resources` directory at the root of the Cordova project. Splash screen dimensions vary for each platform, device and orientation, so a square source image is required the generate each of various sizes. The source image's minimum dimensions should be 2208x2208, and its artwork should be centered within the square, knowning that each generated image will be center cropped into landscape and portait images. The splash screen's artwork should roughly fit within the center 1000x1000 square.

```bash
$ ionic resources --splash
```


### Generating Icons and Splash Screens

To generate both icons and splash screens, follow the instructions above and run:

```bash
$ ionic resources
```

### Platform Specific Resource Images

One source image can be used to generate images for each platform by placing the file within the `resources` directory, such as `resources/icon.png`. To use different source images for individual platforms, place the source image in the respective platform's directory. For example, to use a different icon for Android, it should follow this path: `resources/android/icon.png`, and a different image for iOS would use this path: `resources/ios/icon.png`.


## Update Ionic lib


CHANGELOG.md
### 1.2.14

* Automatically create icon resources from source file: `ionic icon`
* Automatically create splash screen resources from source file: `ionic splash`
* Update config.xml resource platform nodes


### 1.2.13

* Locking Gulp at 3.8.8 to avoid adding the v8flags module dependency


### 1.2.12

* Updating the npm-shrinkwrap


### 1.2.11

* Updating the Labs styles


### 1.2.10

* Updated the serve command for the `serve --labs` to use `IONIC_LAB_URL`


### 1.2.9

* Added proxy-middleware to provide proxying to APIs from the `serve` command
* Updated README doc about how to use the proxy
* Injects platform specific class to HTML to view it as an iOS or Android device in browser
* Introduced Ionic Labs - a way to see preview iOS and Android side by side in the browser
* Bumped `serve-static` to 1.7.1 to avoid some errors with the `serve` for users of Node 0.12
* Added the `add` command to use ionic to manage bower components
* Ionic now reads the Node environment variable `http_proxy` along with the passed `PROXY` variable to get around a local proxy


### 1.2.8

* CSRF cookie fixes


### 1.2.7

* npm-shrinkwrap
* Update ionic.io API URL


### 1.2.6

* Fix `fs.chmodSync: Error: ENOENT` for existing projects
* Fix lib update
* Add ionic app task
* Starter projects can provide `app.json` to specify plugins and sass setup


### 1.2.5

* Do not watch `www/lib/` files by default
* Set watchPatterns within ionic.project config file
* Friendly EMFILE error when too many files are watched
* Ensure config.xml content[src] gets reset after run/emulate
* Improve fetchArchive error handling
* Fix SSL Cert errors
* Do not prompt for address selection when there's only one


### 1.2.4

* Use `cross-spawn` module to fix errors with using spawn on Windows
* Start ionic project from any Github repo
* Start ionic projects using a local directory
* Use specific npm versions in package.json to avoid any future errors from breaking changes
* Fix write errors after downloading github archive files
* Refactor sass setup to use gulpStartupTasks ionic.project property instead


### 1.2.3

* From the server, use `restart` or `r` to restart the client app from the root
* From the server, use `goto` or `g` a url to have the client app navigate to the given url
* From the server, use `consolelogs` or `c` enable/disable console log output
* From the server, use `serverlogs` or `s` to enable/disable server log output
* From the server, use `quit` or `q` to shutdown the server and exit
* Print out Ionic server command tips
* LiveReload server logs specify which device made the request (iOS, Android, etc.)
* Remember address selection [#91](https://github.com/driftyco/ionic-cli/issues/91)
* Reset address selection with `ionic address`
* Add localhost as an option of possible addresses to use [#88](https://github.com/driftyco/ionic-cli/issues/88)
* Inject scripts after charset [#87](https://github.com/driftyco/ionic-cli/issues/87)
* Improved error message when unable to find an IP address [#85](https://github.com/driftyco/ionic-cli/issues/85)
* Fix config.xml errors when in the wrong working directory [#84](https://github.com/driftyco/ionic-cli/issues/84)


### 1.2.2

* ReferenceError hot fix


### 1.2.1

* Clean up any cmd flags which may confuse Cordova [#83](https://github.com/driftyco/ionic-cli/issues/83)
* Select available IP address prompt [#82](https://github.com/driftyco/ionic-cli/issues/82)
* Fix black screen on load [#81](https://github.com/driftyco/ionic-cli/issues/81)


### 1.2.0

* LiveReload from a native app during development
* Option to print out console logs to the terminal
* Option to print out server logs to the terminal
* Start Ionic projects from Codepen urls
* [Live Reload All the Things: Ionic CLI's Latest Features](http://ionicframework.com/blog/live-reload-all-things-ionic-cli/)


lib/ionic.js
/*
 _             _
(_)           (_)
 _  ___  _ __  _  ___
| |/ _ \| '_ \| |/ __|
| | (_) | | | | | (__
|_|\___/|_| |_|_|\___|

http://ionicframework.com/

A utility for starting and administering Ionic based mobile app projects.
Licensed under the MIT license. See LICENSE For more.

Copyright 2014 Drifty (http://drifty.com/)
*/

var argv = require('optimist').argv,
    colors = require('colors'),
    Q = require('q'),
    settings = require('../package.json'),
    shelljs = require('shelljs/global');

colors.setTheme({
  silly: 'rainbow',
  input: 'grey',
  small: 'grey',
  verbose: 'cyan',
  prompt: 'grey',
  info: 'white',
  data: 'grey',
  help: 'cyan',
  warn: 'yellow',
  debug: 'blue',
  error: 'red'
});


var TASKS = [
  {
    title: 'start',
    name: 'start',
    summary: 'Starts a new Ionic project in the specified PATH',
    args: {
      '[options]': 'any flags for the command',
      '<PATH>': 'directory for the new project',
      '[template]': 'Template name, ex: tabs, sidemenu, blank\n' +
                    'Codepen url, ex: http://codepen.io/ionic/pen/odqCz\n' +
                    'Defaults to Ionic "tabs" starter template'
    },
    options: {
      '--appname|-a': 'Human readable name for the app (Use quotes around the name)',
      '--id|-i': 'Package name for <widget id> config, ex: com.mycompany.myapp',
      '--no-cordova|-w': 'Create a basic structure without Cordova requirements',
      '--sass|-s': 'Setup the project to use Sass CSS precompiling'
    },
    module: './ionic/start'
  },
  {
    title: 'serve',
    name: 'serve',
    summary: 'Start a local development server for app dev/testing',
    args: {
      '[options]': ''
    },
    options: {
      '--consolelogs|-c': 'Print app console logs to Ionic CLI',
      '--serverlogs|-s': 'Print dev server logs to Ionic CLI',
      '--port|-p': 'Dev server HTTP port (8100 default)',
      '--livereload-port|-r': 'Live Reload port (35729 default)',
      '--nobrowser|-b': 'Disable launching a browser',
      '--nolivereload': 'Do not start live reload',
      '--noproxy|-x': 'Do not add proxies'
    },
    module: './ionic/serve'
  },
  {
    title: 'platform',
    name: 'platform',
    summary: 'Add platform target for building an Ionic app',
    args: {
      '[options]': '',
      '<PLATFORM>': ''
    },
    module: './ionic/cordova'
  },
  {
    title: 'run',
    name: 'run',
    summary: 'Run an Ionic project on a connected device',
    args: {
      '[options]': '',
      '<PLATFORM>': ''
    },
    options: {
      '--livereload|-l': 'Live reload app dev files from the device' + ' (beta)'.yellow,
      '--port|-p': 'Dev server HTTP port (8100 default, livereload req.)',
      '--livereload-port|-r': 'Live Reload port (35729 default, livereload req.)',
      '--consolelogs|-c': 'Print app console logs to Ionic CLI (livereload req.)',
      '--serverlogs|-s': 'Print dev server logs to Ionic CLI (livereload req.)',
      '--debug|--release': '',
      '--device|--emulator|--target=FOO': ''
    },
    module: './ionic/cordova'
  },
  {
    title: 'emulate',
    name: 'emulate',
    summary: 'Emulate an Ionic project on a simulator or emulator',
    args: {
      '[options]': '',
      '<PLATFORM>': ''
    },
    options: {
      '--livereload|-l': 'Live reload app dev files from the device' + ' (beta)'.yellow,
      '--port|-p': 'Dev server HTTP port (8100 default, livereload req.)',
      '--livereload-port|-r': 'Live Reload port (35729 default, livereload req.)',
      '--consolelogs|-c': 'Print app console logs to Ionic CLI (livereload req.)',
      '--serverlogs|-s': 'Print dev server logs to Ionic CLI (livereload req.)',
      '--debug|--release': '',
      '--device|--emulator|--target=FOO': ''
    },
    module: './ionic/cordova'
  },
  {
    title: 'build',
    name: 'build',
    summary: 'Locally build an Ionic project for a given platform',
    args: {
      '[options]': '',
      '<PLATFORM>': ''
    },
    module: './ionic/cordova'
  },
  {
    title: 'plugin add',
    name: 'plugin',
    summary: 'Add a Cordova plugin',
    args: {
      '[options]': '',
      '<SPEC>': 'Can be a plugin ID, a local path, or a git URL.'
    },
    options: {
      '--searchpath <directory>': 'When looking up plugins by ID, look in this directory\n' +
                                  'and subdirectories first for the plugin before\n' +
                                  'looking it up in the registry.'
    },
    module: './ionic/cordova'
  },
  {
    title: 'prepare',
    name: 'prepare',
    module: './ionic/cordova'
  },
  {
    title: 'compile',
    name: 'compile',
    module: './ionic/cordova'
  },
  {
    title: 'resources',
    name: 'resources',
    summary: 'Automatically create icon and splash screen resources' + ' (beta)'.yellow,
    options: {
      '--icon|-i': 'Generate icon resources',
      '--splash|-s': 'Generate splash screen resources'
    },
    module: './ionic/res/res-build'
  },
  {
    title: 'package',
    name: 'package',
    alt: ['pack'],
    summary: 'Package an app using the Ionic Build service' + ' (beta)'.yellow,
    args: {
      '[options]': '',
      '<MODE>': '"debug" or "release"',
      '<PLATFORM>': '"ios" or "android"'
    },
    options: {
      '--android-keystore-file|-k': 'Android keystore file',
      '--android-keystore-alias|-a': 'Android keystore alias',
      '--android-keystore-password|-w': 'Android keystore password',
      '--android-key-password|-r': 'Android key password',
      '--ios-certificate-file|-c': 'iOS certificate file',
      '--ios-certificate-password|-d': 'iOS certificate password',
      '--ios-profile-file|-f': 'iOS profile file',
      '--output|-o': 'Path to save the packaged app',
      '--no-email|-n': 'Do not send a build package email',
      '--clear-signing|-l': 'Clear out all signing data from Ionic server',
      '--email|-e': 'Ionic account email',
      '--password|-p': 'Ionic account password'
    },
    module: './ionic/package'
  },
  {
    title: 'upload',
    name: 'upload',
    summary: 'Upload an app to your Ionic account',
    options: {
      '--email|-e': 'Ionic account email',
      '--password|-p': 'Ionic account password'
    },
    alt: ['up'],
    module: './ionic/upload'
  },
  {
    title: 'lib',
    name: 'lib',
    summary: 'Gets Ionic library version or updates the Ionic library',
    args: {
      '[options]': '',
      '[update]': 'Updates the Ionic Framework in www/lib/ionic'
    },
    options: {
      '--version|-v': 'Specific Ionic version\nOtherwise it defaults to the latest version'
    },
    module: './ionic/lib'
  },
  {
    title: 'setup',
    name: 'setup',
    summary: 'Configure the project with a build tool ' + '(beta)'.yellow,
    args: {
      '[sass]': 'Setup the project to use Sass CSS precompiling'
    },
    module: './ionic/setup'
  },
  {
    title: 'login',
    name: 'login',
    module: './ionic/login'
  },
  {
    title: 'address',
    name: 'address',
    module: './ionic/serve'
  },
  {
    title: 'app',
    name: 'app',
    // summary: 'Deploy a new Ionic app version or list versions',
    // options: {
    //   '--versions|-v': 'List recently uploaded versions of this app',
    //   '--deploy|-d': 'Upload the current working copy and mark it as deployed',
    //   '--note|-n': 'Add a note to a deploy',
    //   '--uuid|-u': 'Mark an already uploaded version as deployed'
    // },
    module: './ionic/app'
   },
   {
    title: 'service add',
    name: 'service',
    summary: 'Add an Ionic service package and install any required plugins',
    args: {
      '[options]': '',
      '<SPEC>': 'Can be a service name or a git url'
    },
    module: './ionic/service'
   },
   {
    title: 'add',
    name: 'add',
    summary: 'Add a bower component to the project',
    args: {
      'component name': ''
    },
    module: './ionic/add'
   },
   {
    title: 'remove',
    name: 'remove',
    summary: 'Remove a bower component from the project',
    args: {
      'component name': ''
    },
    module: './ionic/add',
    alt: ['rm']
   },
   {
    title: 'list',
    name: 'list',
    summary: 'List bower components from the project',
    args: {
      'component name': ''
    },
    module: './ionic/add',
    alt: ['ls']
   }
];

Ionic = {
  IONIC_DASH: 'https://apps.ionic.io',
  //IONIC_DASH: 'http://localhost:8000',
  IONIC_API: '/api/v1/',
  PRIVATE_PATH: '.ionic',
  _tryBuildingTask: function() {
    if(argv._.length === 0) {
      return false;
    }
    var taskName = argv._[0];

    return this._getTaskWithName(taskName);
  },

  _getTaskWithName: function(name) {
    for(var i = 0; i < TASKS.length; i++) {
      var t = TASKS[i];
      if(t.name === name) {
        return t;
      }
      if(t.alt) {
        for(var j = 0; j < t.alt.length; j++) {
          var alt = t.alt[j];
          if(alt === name) {
            return t;
          }
        }
      }
    }
  },

  printUsage: function(d) {
    var w = function(s) {
      process.stdout.write(s);
    };

    w('\n');

    var rightColumn = 41;
    var dots = '';
    var indent = '';
    var x, arg;

    var taskArgs = d.title;

    for(arg in d.args) {
      taskArgs += ' ' + arg;
    }

    w(taskArgs.green.bold);

    while( (taskArgs + dots).length < rightColumn + 1) {
      dots += '.';
    }

    w(' ' + dots.grey + '  ');

    if(d.summary) {
      w(d.summary.bold);
    }

    for(arg in d.args) {
      if( !d.args[arg] ) continue;

      indent = '';
      w('\n');
      while(indent.length < rightColumn) {
        indent += ' ';
      }
      w( (indent + '    ' + arg + ' ').bold );

      var argDescs = d.args[arg].split('\n');
      var argIndent = indent + '    ';

      for(x=0; x<arg.length + 1; x++) {
        argIndent += ' ';
      }

      for(x=0; x<argDescs.length; x++) {
        if(x===0) {
          w( argDescs[x].bold );
        } else {
          w( '\n' + argIndent + argDescs[x].bold );
        }
      }
    }

    indent = '';
    while(indent.length < d.name.length + 1) {
      indent += ' ';
    }

    var optIndent = indent;
    while(optIndent.length < rightColumn + 4) {
      optIndent += ' ';
    }

    for(var opt in d.options) {
      w('\n');
      dots = '';

      var optLine = indent + '[' + opt + ']  ';

      w(optLine.yellow.bold);

      if(d.options[opt]) {
        while( (dots.length + optLine.length - 2) < rightColumn) {
          dots += '.';
        }
        w(dots.grey + '  ');

        var optDescs = d.options[opt].split('\n');
        for(x=0; x<optDescs.length; x++) {
          if(x===0) {
            w( optDescs[x].bold );
          } else {
            w( '\n' + optIndent + optDescs[x].bold );
          }
        }
      }
    }

    w('\n');
  },

  _printAvailableTasks: function() {
    this._printIonic();
    process.stderr.write('\nUsage: ionic task args\n\n=======================\n\n');

    if(process.argv.length > 2) {
      process.stderr.write( (process.argv[2] + ' is not a valid task\n\n').bold.red );
    }

    process.stderr.write('Available tasks: '.bold);
    process.stderr.write('(use --help or -h for more info)\n\n');

    for(var i = 0; i < TASKS.length; i++) {
      var task = TASKS[i];
      if(task.summary) {
        var name = '   ' + task.name + '  ';
        var dots = '';
        while((name + dots).length < 20) {
          dots += '.';
        }
        process.stderr.write(name.green.bold + dots.grey + '  ' + task.summary.bold + '\n');
      }
    }

    process.stderr.write('\n');
    this.processExit(1);
  },

  _printHelpLines: function() {
    this._printIonic();
    process.stderr.write('\n=======================\n');

    for(var i = 0; i < TASKS.length; i++) {
      var task = TASKS[i];
      if(task.summary) {
        this.printUsage(task);
      }
    }

    process.stderr.write('\n');
    this.processExit(1);
  },

  _printIonic: function() {
    var w = function(s) {
      process.stdout.write(s);
    };

    w('  _             _     \n');
    w(' (_)           (_)     \n');
    w('  _  ___  _ __  _  ___ \n');
    w(' | |/ _ \\| \'_ \\| |/ __|\n');
    w(' | | (_) | | | | | (__ \n');
    w(' |_|\\___/|_| |_|_|\\___|  CLI v'+ settings.version + '\n');

  },

  run: function() {
    var self = this;

    self.checkLatestVersion();

    process.on('exit', function(){
      self.printVersionWarning();
    });

    if( (argv.version || argv.v) && !argv._.length) {
      return this.version();
    }

    if(argv.help || argv.h) {
      return this._printHelpLines();
    }

    if(argv['stats-opt-out']) {
      var IonicStore = require('./ionic/store').IonicStore;
      var ionicConfig = new IonicStore('ionic.config');
      ionicConfig.set('statsOptOut', true);
      ionicConfig.save();
      console.log('Successful stats opt-out');
      return;
    }

    var taskSetting = this._tryBuildingTask();
    if(!taskSetting) {
      return this._printAvailableTasks();
    }
    var taskModule = require(taskSetting.module).IonicTask;

    var taskInstance = new taskModule();
    taskInstance.run(this);
  },

  fail: function(msg, taskHelp) {
    this.hasFailed = true;
    if(msg) {
      process.stderr.write(msg.error.bold);
      process.stderr.write( (' (CLI v' + settings.version + ')').error.bold + '\n');
    }
    if(taskHelp) {
      for(var i = 0; i < TASKS.length; i++) {
        var task = TASKS[i];
        if(task.name == taskHelp) {
          this.printUsage(task);
          process.stderr.write('\n');
          break;
        }
      }
    }
    process.stderr.write('\n');
    this.processExit(1);
  },

  version: function() {
    process.stderr.write(settings.version + '\n');
  },

  checkLatestVersion: function() {
    this.latestVersion = Q.defer();
    var self = this;

    try {
      if(settings.version.indexOf('beta') > -1) {
        // don't bother checking if its a beta
        self.latestVersion.resolve();
        return;
      }

      // stay silent if it errors
      var IonicStore = require('./ionic/store').IonicStore;
      var ionicConfig = new IonicStore('ionic.config');
      var versionCheck = ionicConfig.get('versionCheck');
      if(versionCheck && ((versionCheck + 86400000) > Date.now() )) {
        // we've recently checked for the latest version, so don't bother again
        self.latestVersion.resolve();
        return;
      }

      var proxy = process.env.PROXY || process.env.http_proxy || null;
      var request = require('request');
      request({ url: 'http://registry.npmjs.org/ionic/latest', proxy: proxy }, function(err, res, body) {
        try {
          self.npmVersion = JSON.parse(body).version;
          ionicConfig.set('versionCheck', Date.now());
          ionicConfig.save();
        } catch(e) {}
        self.latestVersion.resolve();
      });
    } catch(e) {
      self.latestVersion.resolve();
    }

  },

  printVersionWarning: function() {
    if(this.npmVersion && this.npmVersion != settings.version.trim()) {
      process.stdout.write('\n-------------------------\n'.red);
      process.stdout.write('Ionic CLI is out of date:\n'.bold.yellow);
      process.stdout.write( (' * Locally installed version: ' + settings.version + '\n').yellow );
      process.stdout.write( (' * Latest version: ' + this.npmVersion + '\n').yellow );
      process.stdout.write( ' * Run '.yellow + 'npm update -g ionic'.bold + ' to update\n'.yellow );
      process.stdout.write('-------------------------\n\n'.red);
      this.npmVersion = null;
    }
  },

  processExit: function(code) {
    this.latestVersion.promise.then(function(){
      process.exit(code);
    });
  },

  getContentSrc: function() {
    var self = this;
    var contentSrc;

    try {
      var fs = require('fs');
      var path = require('path');
      var configXmlPath = path.resolve('config.xml');
      if( !fs.existsSync(configXmlPath) ) {
        return 'index.html';
      }

      self.setConfigXml({
        resetContent: true,
        errorWhenNotFound: false
      });

      var configString = fs.readFileSync(configXmlPath, { encoding: 'utf8' });

      var xml2js = require('xml2js');
      var parseString = xml2js.parseString;
      parseString(configString, function (err, jsonConfig) {
        if(err) {
          return self.fail('Error parsing config.xml: ' + err);
        }
        try {
          contentSrc = jsonConfig.widget.content[0].$.src;
        } catch(e) {
          return self.fail('Error parsing ' + configXmlPath + ': ' + e);
        }
      });

    } catch(e) {
      return self.fail('Error loading ' + configXmlPath + ': ' + e);
    }

    return contentSrc;
  },

  /**
   * Download a zip file, unzip it to a specific folder.
   */
  fetchArchive: function(targetPath, archiveUrl) {
    var os = require('os');
    var fs = require('fs');
    var path = require('path');
    var unzip = require('unzip');
    var q = Q.defer();

    // The folder name the project will be downloaded and extracted to
    console.log('\nDownloading:'.bold, archiveUrl);

    var tmpFolder = os.tmpdir();
    var tempZipFilePath = path.join(tmpFolder, 'ionic-starter-' + new Date().getTime() + '.zip');


    var unzipRepo = function(fileName) {
      var readStream = fs.createReadStream(fileName);
      readStream.on('error', function(err) {
        console.log( ('unzipRepo readStream: ' + err).error );
        q.reject(err);
      });

      var writeStream = unzip.Extract({ path: targetPath });
      writeStream.on('close', function() {
        q.resolve();
      });
      writeStream.on('error', function(err) {
        console.log( ('unzipRepo writeStream: ' + err).error );
        q.reject(err);
      });
      readStream.pipe(writeStream);
    };

    var proxy = process.env.PROXY || process.env.http_proxy || null;
    var request = require('request');
    request({ url: archiveUrl, rejectUnauthorized: false, encoding: null, proxy: proxy }, function(err, res, body) {
      if(err) {
        console.error('Error fetching:'.error.bold, archiveUrl, err);
        q.reject(err);
        return;
      }
      if(!res) {
        console.error('Invalid response:'.error.bold, archiveUrl);
        q.reject('Unable to fetch response: ' + archiveUrl);
        return;
      }
      if(res.statusCode !== 200) {
        if(res.statusCode === 404 || res.statusCode === 406) {
          console.error('Not found:'.error.bold, archiveUrl, '(' + res.statusCode + ')');
          console.error('Please verify the url and try again.'.error.bold);
        } else {
          console.error('Invalid response status:'.error.bold, archiveUrl, '(' + res.statusCode + ')');
        }
        q.reject(res);
        return;
      }
      try {
        fs.writeFileSync(tempZipFilePath, body);
        unzipRepo(tempZipFilePath);
      } catch(e) {
        console.log('fetchArchive request write: ' + e);
        q.reject(e);
      }
    }).on('response', function(res){
      var ProgressBar = require('progress');
      var bar = new ProgressBar('[:bar]  :percent  :etas', {
        complete: '=',
        incomplete: ' ',
        width: 30,
        total: parseInt(res.headers['content-length'], 10)
      });

      res.on('data', function (chunk) {
        try {
          bar.tick(chunk.length);
        } catch(e){}
      });
    });

    return q.promise;
  },

  setConfigXml: function(options) {
    var fs = require('fs');
    var path = require('path');
    var xml2js = require('xml2js');
    var d = Q.defer();

    var self = this;
    var madeChange = false;

    try {
      var configXmlPath = path.resolve('config.xml');

      if(!fs.existsSync(configXmlPath)) {
        // working directory does not have the config.xml file
        if(options.errorWhenNotFound) {
          d.reject('Unable to locate config.xml file. Please ensure the working directory is at the root of the app where the config.xml should be located.');
        } else {
          d.resolve();
        }
        return d.promise;
      }

      var configString = fs.readFileSync(configXmlPath, { encoding: 'utf8' });

      var parseString = xml2js.parseString;
      parseString(configString, function (err, jsonConfig) {
        if(err) {
          d.reject(err);
          return self.fail('Error parsing ' + configXmlPath + ': ' + err);
        }

        if(options.devServer) {
          if( !jsonConfig.widget.content[0].$['original-src'] ) {
            jsonConfig.widget.content[0].$['original-src'] = jsonConfig.widget.content[0].$.src;
            madeChange = true;
          }
          if(jsonConfig.widget.content[0].$.src !== options.devServer) {
            jsonConfig.widget.content[0].$.src = options.devServer;
            madeChange = true;
          }

        } else if(options.resetContent) {

          if( jsonConfig.widget.content[0].$['original-src'] ) {
            jsonConfig.widget.content[0].$.src = jsonConfig.widget.content[0].$['original-src'];
            delete jsonConfig.widget.content[0].$['original-src'];
            madeChange = true;
          }
        }

        if(madeChange) {
          var xmlBuilder = new xml2js.Builder();
          configString = xmlBuilder.buildObject(jsonConfig);
          fs.writeFileSync(configXmlPath, configString);
        }

        d.resolve();
      });

    } catch(e) {
      d.reject(e);
      self.fail('Error updating ' + configXmlPath + ': ' + e);
    }

    return d.promise;
  }

};

var consoleInfo = console.info;
console.info = function() {
  if (arguments.length === 1 && !arguments[0]) return;
  var msg = '';
  for (var n in arguments) {
    msg += arguments[n] + ' ';
  }
  consoleInfo.call(console, msg.blue.bold);
};

var consoleError = console.error;
console.error = function() {
  if (arguments.length === 1 && !arguments[0]) return;
  var msg = ' ✗';
  for (var n in arguments) {
    msg += ' ' + arguments[n];
  }
  consoleError.call(console, msg.red.bold);
};

console.success = function() {
  if (arguments.length === 1 && !arguments[0]) return;
  var msg = ' ✓';
  for (var n in arguments) {
    msg += ' ' + arguments[n];
  }
  console.log(msg.green.bold);
};

exports.Ionic = Ionic;
